relay-profiling/src/sample.rs

```diff
@@ -0,0 +1,429 @@
+
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    is_emulator: Option<bool>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    locale: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    manufacturer: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    model: Option<String>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+enum Version {
+    #[serde(rename = "1")]
+    V1,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct SampleProfile {
+    version: Version,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    debug_meta: Option<DebugMeta>,
+
+    device: DeviceMetadata,
+    os: OSMetadata,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    runtime: Option<RuntimeMetadata>,
+
+    #[serde(default, skip_serializing_if = "String::is_empty")]
+    environment: String,
+    #[serde(alias = "profile_id")]
+    event_id: EventId,
+    platform: Platform,
+    profile: Profile,
+    release: String,
+    timestamp: DateTime<Utc>,
+
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    transactions: Vec<TransactionMetadata>,
+}
+
+impl SampleProfile {
+    fn valid(&self) -> bool {
+        match self.platform {
+            Platform::Cocoa => {
+                self.os.build_number.is_some()
+                    && self.device.is_emulator.is_some()
+                    && self.device.locale.is_some()
+                    && self.device.manufacturer.is_some()
+                    && self.device.model.is_some()
+            }
+            Platform::Python => self.runtime.is_some(),
+            Platform::Node => self.runtime.is_some(),
+            _ => true,
+        }
+    }
+
+    /// Removes a sample when it's the only sample on its thread
+    fn remove_single_samples_per_thread(&mut self) {
+        let mut sample_count_by_thread_id: HashMap<u64, u32> = HashMap::new();
+
+        for sample in &self.profile.samples {
+            *sample_count_by_thread_id
+                .entry(sample.thread_id)
+                .or_default() += 1;
+        }
+
+        // Only keep data from threads with more than 1 sample so we can calculate a duration
+        sample_count_by_thread_id.retain(|_, count| *count > 1);
+
+        self.profile
+            .samples
+            .retain(|sample| sample_count_by_thread_id.contains_key(&sample.thread_id));
+    }
+
+    fn strip_pointer_authentication_code(&mut self) {
+        self.profile
+            .strip_pointer_authentication_code(&self.platform, &self.device.architecture);
+    }
+}
+
+pub fn expand_sample_profile(payload: &[u8]) -> Result<Vec<Vec<u8>>, ProfileError> {
+    let profile = parse_profile(payload)?;
+
+    if !profile.valid() {
+        return Err(ProfileError::MissingProfileMetadata);
+    }
+
+    let mut items: Vec<Vec<u8>> = Vec::new();
+
+    for transaction in &profile.transactions {
+        let mut new_profile = profile.clone();
```

Comment:
What is the purpose of splitting each transaction into a separate profile? A doc comment would help on this function.

Reply:
I left a comment, let me know if it's clear.

Reply:
Yes, thank you 👍